fdir.1
.TH FDir 1 local
.SH NAME
fdir.sh \- favorite directory.
.SH SYNOPSIS
.B fdir.sh
.RB [ " init " ]
.br
.B fdir.sh
.RB [ " \-l " ]
.br
.B fdir.sh
.RB [ " \-v " ]
.br
.B fdir.sh
.RB [ " \-h " ]
.br
.ll +8
.B fdir.sh
.RB [ " \-s " ]
[
.I "key"
]
.br
.ll +8
.B fdir.sh
.RB [ " \-r " ]
[
.I "key"
]
.SH DESCRIPTION
.I FDir
is a cli directory shortcut manager.
.br
It requires you to
.B cd
to target directory first in order to create a shortcut.
.PP
.I FDir
let you use
.B key
to refer to a directory.
.br
It works by save all your predefine
.B key 
to 
.I ~/.fdirrc 
and make 
.B alias
.br
You can access all directory shortcuts by simply 
.B source 
.I ~/.fdirrc
.SH TIPS
It recommends to alias 
.I fdir.sh 
to 'fdir'or 'fd'
.br
Although, this is depends on your 
.B PREFIX_
in
.I fdir.sh
config section.
.SH USAGE
Assume you have already 
.B alias
.I fdir.sh
to 'fdir'
.PP
.B Add Favorite Directory
.PP
cd /path/to/dir
.br
fdir -s key
.PP
.B Remove Favorite Directory
.PP
fdir -r key
.PP
.B Change Directory to your Favorite Directory
.br
<prefix><key>
.br
Assume your 
.B PREFIX_ 
is 'fd-' and
.B key
is 'mykey'
.PP
fd-mykey
.SH OPTIONS
.TP
.B init
Automatic create ~/.fdirrc and add 'source ~/.fdirrc' to your shell init rc.
Default is '.bashrc'.
.TP
.B \-l
List favorite directory and key.
.TP
.B \-v
Show version.
.TP
.B \-h
Show help.
.TP
.B \-s
Add key.
.TP
.B \-r
Remove key.
.SH CONFIG
You can edit config by open 
.I fdir.sh
file in 
.B USER CONFIG 
section.
.PP
.B PREFIX_ 
= prefix of key. default is 'fd-'
.br
.B FILENAME_ 
= data file name. default is '.fdirrc'
.br
.B SAVE_FLAG_ 
= save flag use to save key. default is '-s'
.br
.B REMOVE_FLAG_ 
= remove flag use to remove key. default is '-r'
.br
.B LIST_FLAG_ 
= list flag use to list all key. default is '-l'
.SH  ISSUES
.I FDir
requires to restart terminal after
.I <Add/Remove>
.B key
.br
If you change config
.B FILENAME_
,You must edit your init shell rc. (default is '.bashrc')
.SH  AUTHOR
.br
Program : RealtimeBagIdea
.br
Manpage : CSaratakij
.SH REPORTING BUGS
Report bugs to <https://github.com/RealtimeBagIdea/FDir/issues>.
.SH LICENSE
.I GNU General Public License v3.0
.SH LINKS
.I Git :
git clone https://github.com/RealtimeBagIdea/FDir.git


fdir.sh
#!/bin/bash

#-------------- How to use -----------------#
# 1 move this file to /usr/bin
# 2 run 'sudo chmod 777 /usr/bin/fdir.sh'
# 3 run 'fdir.sh init'
# 4 re-open terminal
#-------------------------------------------#

#-------------------------------- Update -------------------------------------------------------------------#
#-------+---------------------------------------------------------------------------------------------------#
#  Ver  |   Detail                                                                                          |
#-------+---------------------------------------------------------------------------------------------------#
#  1.1  |   Add Prefix Editor                                                                               |
#-------+---------------------------------------------------------------------------------------------------#
#  1.2  |   Add Init                                                                                        |
#-------+---------------------------------------------------------------------------------------------------#


#=============================== USER CONFIG ===============================================================#

PREFIX_="fd-"
SAVE_FLAG_="-s"
REMOVE_FLAG_="-r"
LIST_FLAG_="-l"
FILENAME_=".fdirrc" #If you change this value. You must edit your .bashrc

#================================ END CONFIG ===============================================================#


#Variable
VERSION="1.2"
CURRENTUSER=$(whoami)

#==================== SETUP ========================#

if [ "$CURRENTUSER" == "root" ]; then
    FD_HOME="/$CURRENTUSER"
else
    FD_HOME="/home/$CURRENTUSER"
fi
CONFIG_FILE=$FD_HOME/$FILENAME_

if [ ! -f $CONFIG_FILE ]; then
    echo "" > $CONFIG_FILE
fi

#==================== METHOD =======================#
Init ()
{
    read -p "Enter your shell init file (default: .bashrc) : " SHELLINIT 

    if [[ $SHELLINIT = "" ]]; then
        SHELLINIT=".bashrc"
    fi

    SHELLINIT="$FD_HOME/$SHELLINIT"

    IS_INITED=$(cat $SHELLINIT | grep "source $CONFIG_FILE")

    if [ "$IS_INITED" == "" ]; then
        echo "source $CONFIG_FILE" >> "$SHELLINIT"
        echo "Init successfully (You have to restart Terminal)"
    else
        echo "FDir init already"
    fi
    echo ""
}

ShowError() 
{
    echo "Command is not correct, see fdir.sh -h"
    echo ""
}

Save()
{
    if [ "$1" != "" ]; then

        IS_EXISTS=$(cat $CONFIG_FILE | grep "alias $PREFIX_$1=")

        if [ "$IS_EXISTS" == "" ]; then
            CORRECT_PATH=$(echo $PWD | sed "s/\s/\\\ /g")
            echo "alias $PREFIX_$1=\"cd $CORRECT_PATH\"" >> $CONFIG_FILE
            echo "Add '$1' to $PWD"
            echo ""
        else
            echo "Key '$1' is exist"
            echo ""
        fi
    else
        ShowError
    fi
}

Remove()
{
    if [ "$1" != "" ]; then

        IS_EXISTS=$(cat $CONFIG_FILE | grep "alias $PREFIX_$1=")

        if [ "$IS_EXISTS" == "" ]; then
            echo "Key '$1' not found"
            echo ""
        else
            sed -i "/alias $PREFIX_$1=.*$/d" $CONFIG_FILE
            echo "Remove '$1' successfully"
            echo ""
        fi
    else
        ShowError
    fi
}

List()
{
    FILE=$(cat $CONFIG_FILE | sed -e "s/alias $PREFIX_//g")
    FILE=$(echo $FILE | sed -e "s/=\"cd /#TAB#=#SPACE#/g")
    FILE=$(echo $FILE | sed -e "s/\"//g")
    FILE=$(echo $FILE | sed -e "s/\\\ /#SPACE#/g")

    for i in ${FILE[@]}
    do
        DECODE=$(echo $i | sed -e "s/#SPACE#/ /g")
        DECODE=$(echo $DECODE | sed -e "s/#TAB#/\\t/g")
        echo -e "$DECODE"
    done
    echo ""
}

Help()
{
    echo -e "Use for Edit: \t fdir.sh [OPTION]"
    echo -e "Use for Move: \t <prefix><key> (Prefix can setting in this script. Default is fd-)"
    echo ""
    echo -e "[INIT]"
    echo -e "\tinit \tCreate '.fdirrc' file and add 'source ~/.fdirrc' to shell init file"
    echo ""
    echo -e "[OPTION]"
    echo -e "\t-s <key> \tLink <key> to current directory"
    echo -e "\t-r <key> \tRemove <key>"
    echo -e "\t-l \t\tList all <key>"
    echo -e "\t-v \t\tShow version"
    echo ""
    echo -e "[EXAMPLE]"
    echo -e "\t#Terminal 1"
    echo -e "\t\t$ cd /path/to/dir"
    echo -e "\t\t$ fdir.sh -s mydir"
    echo -e "\t#Terminal 2"
    echo -e "\t\t$ fd-mydir"
}


#==================== MAIN =======================#

if [ "$1" == "$SAVE_FLAG_" ]; then
    Save $2
elif [ "$1" == "$REMOVE_FLAG_" ]; then
    Remove $2
elif [ "$1" == "$LIST_FLAG_" ]; then
    List   
elif [ "$1" == "-h" ]; then
    Help
elif [ "$1" == "-v" ]; then
    echo "Version: $VERSION" 
    echo ""
elif [ "$1" == "init" ]; then
    Init
else
    ShowError
fi
